File: scripts/selver_api.py
"""
Selver.ee API Client
Vue Storefront + Magento backend
Index: vue_storefront_catalog_et
"""

import json
import os
import subprocess
import sys
import urllib.parse
import requests

TOKEN_FILE = os.path.join(os.path.dirname(__file__), "../data/selver_token.json")

BASE = "https://www.selver.ee"
INDEX = "vue_storefront_catalog_et"
STORE = "et"

HEADERS = {
    "Accept": "application/json",
    "Content-Type": "application/json",
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
}

# Known category IDs from Selver catalog
CATEGORIES = {
    "piim_koor": 234,         # Piimad, koored
    "kohupiim": 235,          # Kohupiimad, kodujuustud
    "jogurt": 236,            # Jogurtid
    "munad": 239,             # Munad
    "void_margariinid": 240,  # Võid, margariinid
    "juustud": 243,           # Juustud
    "leivad": 248,            # Leivad
    "sealiha": 219,           # Sealiha
    "kana": 220,              # (likely near 219)
    "kala": 221,              # (likely)
    "salatid": 257,           # Salatid
    "maitseained": 263,       # Maitseained
    "olid": 267,              # Õlid, äädikad
    "kuivained": 9,           # Kuivained, hommikusöögid
    "joogid": 48,             # Veed, mahlad, joogid
    "kulmutatud": 285,        # Külmutatud liha- ja kalatooted
    "konservid": None,        # TBD
}


class SelverAPI:
    def __init__(self, token: str = None):
        self.token = token
        self.cart_token = None
        self.cookies = {}
        self.session = requests.Session()
        self.session.headers.update(HEADERS)

    def _auth_headers(self) -> dict:
        """Return Authorization + Cookie headers if token/cookies are set."""
        hdrs = {}
        if self.token:
            hdrs["Authorization"] = f"Bearer {self.token}"
        if self.cookies:
            hdrs["Cookie"] = "; ".join(f"{k}={v}" for k, v in self.cookies.items())
        return hdrs

    def _get(self, path: str, params: dict = None) -> dict:
        """GET with auth header and storeCode."""
        p = {"storeCode": STORE, **(params or {})}
        r = self.session.get(f"{BASE}{path}", params=p, headers=self._auth_headers())
        return r.json()

    def _post(self, path: str, data: dict = None, params: dict = None,
              extra_params: dict = None) -> dict:
        """POST with auth header and storeCode."""
        p = {"storeCode": STORE, **(params or {}), **(extra_params or {})}
        r = self.session.post(f"{BASE}{path}", json=data or {}, params=p, headers=self._auth_headers())
        return r.json()

    # ── AUTH ──────────────────────────────────────────────────────────────────

    def load_token(self) -> bool:
        """Load saved token from file. Returns True if token found."""
        if not os.path.exists(TOKEN_FILE):
            return False
        try:
            with open(TOKEN_FILE) as f:
                data = json.load(f)
            token = data.get("token")
            if token:
                if isinstance(token, str) and token.startswith('"') and token.endswith('"'):
                    token = token[1:-1]
                self.token = token
                self.cart_token = data.get("cart_token")
                self.cookies = data.get("cookies", {})
                return True
        except Exception:
            pass
        return False

    def login_browser(self, id_code: str = "38511080251", timeout: int = 120) -> bool:
        """
        Launch browser-based Smart-ID login flow.
        Prints verification code for user to confirm.
        Saves token to data/selver_token.json.
        Returns True on success.
        """
        script = os.path.join(os.path.dirname(__file__), "selver_auth.py")
        result = subprocess.run(
            [sys.executable, script, "--id-code", id_code, "--timeout", str(timeout)],
            capture_output=False,  # let output print directly
        )
        if result.returncode == 0:
            return self.load_token()
        return False

    def ensure_auth(self, id_code: str = "38511080251") -> bool:
        """Load token from file, or trigger browser login if missing/expired."""
        if self.token:
            return True
        if self.load_token():
            # Quick check if token still valid
            try:
                r = self.session.get(f"{BASE}/api/user/me?token={self.token}", timeout=5)
                if r.status_code == 200 and r.json().get("code") == 200:
                    return True
            except Exception:
                pass
        # Token missing or expired — do browser login
        print("Token missing or expired. Starting browser login...")
        return self.login_browser(id_code=id_code)

    def get_profile(self) -> dict:
        return self._get("/api/user/me")

    # ── CATALOG ───────────────────────────────────────────────────────────────

    def _catalog_search(self, entity: str, query: dict, size: int = 20,
                        from_: int = 0, sort: str = "") -> dict:
        encoded = urllib.parse.quote(json.dumps(query))
        url = (f"{BASE}/api/catalog/{INDEX}/{entity}/_search"
               f"?from={from_}&size={size}&sort={sort}&request={encoded}")
        resp = self.session.get(url)
        return resp.json()

    def search_products(self, text: str = None, category_id: int = None,
                        size: int = 20) -> list:
        """Search products by text (wildcard) and/or category_id"""
        must = [{"terms": {"status": [0, 1]}}]

        if category_id:
            must.append({"terms": {"category_ids": [category_id]}})

        if text:
            # Wildcard search on name - works with Estonian text
            words = text.lower().split()
            for word in words:
                must.append({"wildcard": {"name": f"*{word}*"}})

        query = {"query": {"bool": {"filter": {"bool": {"must": must}}}}}
        result = self._catalog_search("product", query, size=size)
        hits = result.get("hits", {}).get("hits", [])
        return [self._format_product(h["_source"]) for h in hits]

    def get_product_by_sku(self, sku: str) -> dict | None:
        query = {"query": {"bool": {"filter": {"terms": {"sku": [sku]}}}}}
        result = self._catalog_search("product", query, size=1)
        hits = result.get("hits", {}).get("hits", [])
        if hits:
            return self._format_product(hits[0]["_source"])
        return None

    def get_categories(self, level: int = 4) -> list:
        query = {
            "query": {
                "bool": {
                    "filter": {
                        "bool": {
                            "must": [
                                {"terms": {"level": [level]}},
                                {"terms": {"is_active": [True]}},
                            ]
                        }
                    }
                }
            }
        }
        result = self._catalog_search("category", query, size=200, sort="position:asc")
        hits = result.get("hits", {}).get("hits", [])
        return [
            {
                "id": h["_source"]["id"],
                "name": h["_source"]["name"],
                "url_path": h["_source"].get("url_path"),
            }
            for h in hits
        ]

    def _format_product(self, src: dict) -> dict:
        stock = src.get("stock", {})
        return {
            "sku": src.get("sku"),
            "name": src.get("name"),
            "price": round(src.get("price", 0), 2),
            "special_price": src.get("special_price"),
            "unit": src.get("product_sales_unit"),
            "brand": src.get("product_brand"),
            "category_ids": src.get("category_ids", []),
            "product_id": stock.get("product_id"),
            "stock_status": stock.get("stock_status"),  # 1 = available
        }

    # ── CART ──────────────────────────────────────────────────────────────────

    def create_cart(self) -> str:
        """Create a new cart, returns cart quote ID"""
        result = self._post("/api/cart/create")
        cart_id = result.get("result")
        if cart_id:
            self.cart_token = str(cart_id)
        return cart_id

    def get_cart(self, cart_id: str = None) -> dict:
        cart_id = cart_id or self.cart_token
        return self._get("/api/cart/pull", {"cartId": cart_id})

    def add_to_cart(self, sku: str, qty: float = 1.0, product_id: int = None,
                    cart_id: str = None) -> dict:
        cart_id = cart_id or self.cart_token
        item = {"sku": sku, "qty": qty}
        if product_id:
            item["product_id"] = product_id
        return self._post(
            "/api/ext/tkg-sales/cart/add-cart-items",
            {"cartId": cart_id, "cartItems": [item]},
        )

    def add_multiple_to_cart(self, items: list, cart_id: str = None) -> dict:
        """items: [{"sku": "...", "qty": 1, "product_id": ...}, ...]"""
        cart_id = cart_id or self.cart_token
        return self._post(
            "/api/ext/tkg-sales/cart/add-cart-items",
            {"cartId": cart_id, "cartItems": items},
        )

    def clear_cart(self, cart_id: str = None) -> dict:
        cart_id = cart_id or self.cart_token
        return self._post(
            "/api/ext/tkg-sales/cart/truncate",
            {"cartId": cart_id},
        )

    def get_cart_totals(self, cart_id: str = None) -> dict:
        cart_id = cart_id or self.cart_token
        return self._get("/api/cart/totals", {"cartId": cart_id})

    def get_shipping_methods(self, address: dict, cart_id: str = None) -> dict:
        cart_id = cart_id or self.cart_token
        return self._post(
            "/api/cart/shipping-methods",
            {"cartId": cart_id, "address": address},
        )

    # ── TIMESLOTS ─────────────────────────────────────────────────────────────

    # Delivery method ID for courier (Tellin kauba kulleriga)
    DELIVERY_METHOD_ID = 419

    def validate_postcode(self, postcode: str) -> dict:
        return self._get(f"/api/ext/timeslot/is-postcode-valid/{postcode}")

    def get_timeslot_reservation(self, cart_id: str = None) -> dict:
        """Get current active timeslot reservation."""
        cart_id = cart_id or self.cart_token
        return self._get("/api/ext/timeslot/get-reservation", {"cartId": cart_id})

    def get_available_timeslots(self, method_id: int = None, cart_id: str = None) -> dict:
        """
        Get available delivery timeslots.
        Returns: {days: [...], times: [...], slots: [{id, date, time, label, enabled}, ...]}
        Only slots with enabled=True are available for booking.
        """
        method_id = method_id or self.DELIVERY_METHOD_ID
        cart_id = cart_id or self.cart_token
        return self._get(f"/api/ext/timeslot/get-method-data/{method_id}/0",
                         {"cartId": cart_id})

    def reserve_timeslot(self, timeslot_id: int, date: str,
                         email: str = None, telephone: str = None,
                         cart_id: str = None) -> dict:
        """
        Reserve a delivery timeslot.
        timeslot_id: slot id from get_available_timeslots()
        date: "YYYY-MM-DD"
        Returns reservation info with expired_at timestamp (45 min window).
        """
        cart_id = cart_id or self.cart_token
        body = {
            "timeslotId": timeslot_id,
            "date": date,
            "stationId": "",
            "email": email or "",
            "telephone": telephone or "",
        }
        r = self.session.post(
            f"{BASE}/api/ext/timeslot/reserve",
            json=body,
            params={"cartId": cart_id, "storeCode": STORE},
            headers=self._auth_headers(),
        )
        return r.json()

    def delete_timeslot_reservation(self, cart_id: str = None) -> dict:
        """Cancel current timeslot reservation."""
        cart_id = cart_id or self.cart_token
        r = self.session.post(
            f"{BASE}/api/ext/timeslot/delete",
            json={},
            params={"cartId": cart_id, "storeCode": STORE},
            headers=self._auth_headers(),
        )
        return r.json()

    def get_shipping_methods_for_address(self, address: dict = None,
                                          cart_id: str = None) -> list:
        """
        Get available shipping methods for given address.
        Default address is Ivan's home (Eha 3, Taebla).
        Returns list of {carrier_code, method_code, method_title, price_incl_tax}.
        """
        if address is None:
            address = {
                "region": "Laane maakond", "region_id": 0, "country_id": "EE",
                "street": ["Eha tanav 3", "Taebla"], "city": "Taebla",
                "postcode": "90801", "firstname": "Ivan", "lastname": "Sokolov",
                "email": "0959741@gmail.com", "telephone": "37259025467",
                "extension_attributes": {},
            }
        cart_id = cart_id or self.cart_token
        result = self._post("/api/cart/shipping-methods",
                            {"address": address},
                            params={"cartId": cart_id})
        return result.get("result", []) if result.get("code") == 200 else []


if __name__ == "__main__":
    api = SelverAPI()
    print("Testing product search...")
    products = api.search_products(text="täispiim", size=5)
    for p in products:
        print(f"  {p['name']} | {p['sku']} | {p['price']}€")
    print(f"\nFound {len(products)} products\n")

    print("Category 234 (Piimad):")
    products = api.search_products(category_id=234, size=5)
    for p in products:
        print(f"  {p['name']} | {p['price']}€")

File: mcp-server/selver-api.ts
/**
 * Selver.ee API client.
 * Vue Storefront + Magento backend.
 * Catalog index: vue_storefront_catalog_et
 */

import { readFileSync, writeFileSync, existsSync } from 'fs'
import { join, dirname } from 'path'

const BASE = 'https://www.selver.ee'
const INDEX = 'vue_storefront_catalog_et'
const STORE = 'et'

const UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'

const DATA_DIR = join(dirname(import.meta.dir), 'data')
const TOKEN_FILE = join(DATA_DIR, 'selver_token.json')
const PANTRY_FILE = join(DATA_DIR, 'pantry.json')
const CONFIG_FILE = join(DATA_DIR, 'config.json')

// ── Types ──────────────────────────────────────────────────────────────────

export type Product = {
  sku: string
  name: string
  price: number
  special_price: number | null
  unit: string | null
  brand: string | null
  category_ids: number[]
  product_id: number | null
  stock_status: number | null
  url_path?: string
  image?: string
}

export type CartItem = {
  item_id: number
  sku: string
  qty: number
  name: string
  price: number
  product_type: string
}

export type TokenData = {
  token: string | null
  cart_token: string | null
  cookies: Record<string, string>
  user?: Record<string, any>
  id_code?: string
  saved_at?: number
}

export type PantryItem = {
  name: string
  quantity?: number
  unit?: string
  category?: string
  image?: string
  added?: string
}

// ── Known categories ───────────────────────────────────────────────────────

export const CATEGORIES: Record<string, { id: number; name_et: string }> = {
  milk:        { id: 234, name_et: 'Piimad, koored' },
  cottage:     { id: 235, name_et: 'Kohupiimad, kodujuustud' },
  yogurt:      { id: 236, name_et: 'Jogurtid' },
  eggs:        { id: 239, name_et: 'Munad' },
  butter:      { id: 240, name_et: 'Võid, margariinid' },
  cheese:      { id: 243, name_et: 'Juustud' },
  bread:       { id: 248, name_et: 'Leivad' },
  pork:        { id: 219, name_et: 'Sealiha' },
  chicken:     { id: 220, name_et: 'Linnuliha' },
  fish:        { id: 221, name_et: 'Kala' },
  salads:      { id: 257, name_et: 'Salatid' },
  spices:      { id: 263, name_et: 'Maitseained' },
  oils:        { id: 267, name_et: 'Õlid, äädikad' },
  dry_goods:   { id: 9,   name_et: 'Kuivained, hommikusöögid' },
  drinks:      { id: 48,  name_et: 'Veed, mahlad, joogid' },
  frozen_meat: { id: 285, name_et: 'Külmutatud liha- ja kalatooted' },
  fruits:      { id: 253, name_et: 'Puuviljad' },
  vegetables:  { id: 254, name_et: 'Köögiviljad' },
  sausages:    { id: 225, name_et: 'Vorstid, viinerid' },
  sweets:      { id: 270, name_et: 'Maiustused' },
  baby:        { id: 296, name_et: 'Lastekaubad' },
  pets:        { id: 303, name_et: 'Lemmikloomakaubad' },
  household:   { id: 289, name_et: 'Majapidamistarbed' },
}

// ── API Client ─────────────────────────────────────────────────────────────

export class SelverAPI {
  token: string | null = null
  cartToken: string | null = null
  cookies: Record<string, string> = {}
  user: Record<string, any> = {}

  constructor() {
    this.loadToken()
  }

  // ── Auth ────────────────────────────────────────────────────────────────

  loadToken(): boolean {
    if (!existsSync(TOKEN_FILE)) return false
    try {
      const data: TokenData = JSON.parse(readFileSync(TOKEN_FILE, 'utf8'))
      let token = data.token
      if (token?.startsWith('"') && token.endsWith('"')) {
        token = token.slice(1, -1)
      }
      this.token = token
      this.cartToken = data.cart_token
      this.cookies = data.cookies || {}
      this.user = data.user || {}
      return !!token
    } catch {
      return false
    }
  }

  private authHeaders(): Record<string, string> {
    const hdrs: Record<string, string> = {
      'Accept': 'application/json',
      'Content-Type': 'application/json',
      'User-Agent': UA,
    }
    if (this.token) {
      hdrs['Authorization'] = `Bearer ${this.token}`
    }
    if (Object.keys(this.cookies).length > 0) {
      hdrs['Cookie'] = Object.entries(this.cookies).map(([k, v]) => `${k}=${v}`).join('; ')
    }
    return hdrs
  }

  /** Check if current token is valid */
  async checkAuth(): Promise<{ valid: boolean; user?: string; error?: string; expires_in?: number }> {
    if (!this.token) return { valid: false, error: 'No token loaded' }

    // Decode JWT to check expiry
    try {
      const payload = JSON.parse(Buffer.from(this.token.split('.')[1], 'base64url').toString())
      const now = Math.floor(Date.now() / 1000)
      if (payload.exp && now >= payload.exp) {
        return { valid: false, error: `Token expired ${Math.floor((now - payload.exp) / 60)} minutes ago` }
      }
      const expiresIn = payload.exp ? payload.exp - now : undefined

      // Verify with server
      const r = await fetch(`${BASE}/api/user/me?token=${this.token}`, { headers: this.authHeaders() })
      const body = await r.json() as any
      if (body.code === 200) {
        return {
          valid: true,
          user: `${body.result.firstname} ${body.result.lastname}`,
          expires_in: expiresIn,
        }
      }
      return { valid: false, error: body.result || 'Unknown error' }
    } catch (e: any) {
      return { valid: false, error: e.message }
    }
  }

  /** Trigger browser-based Smart-ID login (requires user interaction) */
  async triggerAuth(idCode: string = '38511080251', timeout: number = 180): Promise<string> {
    const { execSync } = await import('child_process')
    const script = join(dirname(import.meta.dir), 'scripts', 'selver_auth.py')
    try {
      const output = execSync(
        `python3 ${script} --id-code ${idCode} --timeout ${timeout}`,
        { timeout: (timeout + 30) * 1000, encoding: 'utf8', stdio: ['pipe', 'pipe', 'pipe'] }
      )
      this.loadToken()
      // Extract verification code from output
      const codeMatch = output.match(/Verification code: (\d{4})/)
      const code = codeMatch ? codeMatch[1] : null
      return code
        ? `Smart-ID verification code: ${code}. Confirm on your phone!`
        : output
    } catch (e: any) {
      return `Auth failed: ${e.stderr || e.message}`
    }
  }

  // ── Catalog ─────────────────────────────────────────────────────────────

  private async catalogSearch(entity: string, query: object, size = 20, from = 0, sort = ''): Promise<any> {
    const encoded = encodeURIComponent(JSON.stringify(query))
    const url = `${BASE}/api/catalog/${INDEX}/${entity}/_search?from=${from}&size=${size}&sort=${sort}&request=${encoded}`
    const r = await fetch(url, { headers: { 'User-Agent': UA } })
    return r.json()
  }

  private formatProduct(src: any): Product {
    const stock = src.stock || {}
    return {
      sku: src.sku,
      name: src.name,
      price: Math.round((src.price || 0) * 100) / 100,
      special_price: src.special_price ? Math.round(src.special_price * 100) / 100 : null,
      unit: src.product_sales_unit || null,
      brand: src.product_brand || null,
      category_ids: src.category_ids || [],
      product_id: stock.product_id || null,
      stock_status: stock.stock_status || null,
      url_path: src.url_path || null,
      image: src.image || null,
    }
  }

  /** Search products by text and/or category */
  async searchProducts(text?: string, categoryId?: number, size = 20): Promise<Product[]> {
    const must: any[] = [{ terms: { status: [0, 1] } }]

    if (categoryId) {
      must.push({ terms: { category_ids: [categoryId] } })
    }
    if (text) {
      for (const word of text.toLowerCase().split(/\s+/)) {
        must.push({ wildcard: { name: `*${word}*` } })
      }
    }

    const query = { query: { bool: { filter: { bool: { must } } } } }
    const result = await this.catalogSearch('product', query, size)
    const hits = result?.hits?.hits || []
    return hits.map((h: any) => this.formatProduct(h._source))
  }

  /** Get product by SKU */
  async getProduct(sku: string): Promise<Product | null> {
    const query = { query: { bool: { filter: { terms: { sku: [sku] } } } } }
    const result = await this.catalogSearch('product', query, 1)
    const hits = result?.hits?.hits || []
    return hits.length ? this.formatProduct(hits[0]._source) : null
  }

  /** Browse category tree */
  async getCategories(level = 4): Promise<{ id: number; name: string; url_path: string }[]> {
    const query = {
      query: { bool: { filter: { bool: { must: [
        { terms: { level: [level] } },
        { terms: { is_active: [true] } },
      ] } } } }
    }
    const result = await this.catalogSearch('category', query, 200, 0, 'position:asc')
    const hits = result?.hits?.hits || []
    return hits.map((h: any) => ({
      id: h._source.id,
      name: h._source.name,
      url_path: h._source.url_path,
    }))
  }

  /** Get products on sale (special_price set and lower than price) */
  async getDeals(categoryId?: number, size = 20): Promise<Product[]> {
    const must: any[] = [
      { terms: { status: [0, 1] } },
      { exists: { field: 'special_price' } },
    ]
    if (categoryId) {
      must.push({ terms: { category_ids: [categoryId] } })
    }

    const query = { query: { bool: { filter: { bool: { must } } } } }
    const result = await this.catalogSearch('product', query, size, 0, 'special_price:asc')
    const hits = result?.hits?.hits || []
    return hits
      .map((h: any) => this.formatProduct(h._source))
      .filter(p => p.special_price && p.special_price < p.price)
  }

  // ── Cart ────────────────────────────────────────────────────────────────

  private async apiGet(path: string, params: Record<string, string> = {}): Promise<any> {
    const p = new URLSearchParams({ storeCode: STORE, ...params })
    const r = await fetch(`${BASE}${path}?${p}`, { headers: this.authHeaders() })
    return r.json()
  }

  private async apiPost(path: string, body: any = {}, params: Record<string, string> = {}): Promise<any> {
    const p = new URLSearchParams({ storeCode: STORE, ...params })
    const r = await fetch(`${BASE}${path}?${p}`, {
      method: 'POST',
      headers: this.authHeaders(),
      body: JSON.stringify(body),
    })
    return r.json()
  }

  /** Create a new cart, returns cart ID */
  async createCart(): Promise<string | null> {
    const result = await this.apiPost('/api/cart/create')
    const cartId = result?.result
    if (cartId) {
      this.cartToken = String(cartId)
      // Save to token file
      if (existsSync(TOKEN_FILE)) {
        const data = JSON.parse(readFileSync(TOKEN_FILE, 'utf8'))
        data.cart_token = this.cartToken
        writeFileSync(TOKEN_FILE, JSON.stringify(data, null, 2))
      }
    }
    return cartId
  }

  /** Get current cart contents */
  async getCart(): Promise<{ items: CartItem[]; error?: string }> {
    if (!this.cartToken) {
      return { items: [], error: 'No cart token. Create a cart or authenticate first.' }
    }
    const result = await this.apiGet('/api/cart/pull', { cartId: this.cartToken })
    if (result?.code === 200) {
      return { items: result.result || [] }
    }
    return { items: [], error: result?.result || 'Failed to fetch cart' }
  }

  /** Add items to cart. cartId goes in query params (not body!) */
  async addToCart(items: { sku: string; qty: number }[]): Promise<{ success: boolean; cart?: any; error?: string }> {
    if (!this.cartToken) {
      return { success: false, error: 'No cart token. Authenticate first.' }
    }
    const result = await this.apiPost(
      '/api/ext/tkg-sales/cart/add-cart-items',
      { cartItems: items },
      { cartId: this.cartToken },
    )
    if (result?.code === 200) {
      return { success: true, cart: result.result }
    }
    return { success: false, error: JSON.stringify(result) }
  }

  /** Remove item from cart by item_id */
  async removeFromCart(itemId: number): Promise<{ success: boolean; error?: string }> {
    if (!this.cartToken) {
      return { success: false, error: 'No cart token.' }
    }
    const result = await this.apiPost(
      '/api/cart/delete',
      { cartItem: { item_id: itemId } },
      { cartId: this.cartToken },
    )
    if (result?.code === 200) {
      return { success: true }
    }
    return { success: false, error: JSON.stringify(result) }
  }

  /** Update item quantity */
  async updateCartItem(itemId: number, sku: string, qty: number): Promise<{ success: boolean; error?: string }> {
    if (!this.cartToken) {
      return { success: false, error: 'No cart token.' }
    }
    const result = await this.apiPost(
      '/api/cart/update',
      { cartItem: { item_id: itemId, sku, qty, quote_id: this.cartToken } },
      { cartId: this.cartToken },
    )
    if (result?.code === 200) {
      return { success: true }
    }
    return { success: false, error: JSON.stringify(result) }
  }

  /** Clear all items from cart */
  async clearCart(): Promise<{ success: boolean; error?: string }> {
    if (!this.cartToken) {
      return { success: false, error: 'No cart token.' }
    }
    const result = await this.apiPost(
      '/api/ext/tkg-sales/cart/truncate',
      {},
      { cartId: this.cartToken },
    )
    if (result?.code === 200) {
      return { success: true }
    }
    return { success: false, error: JSON.stringify(result) }
  }

  /** Get cart totals (subtotal, grand total, item count) */
  async getCartTotals(): Promise<any> {
    if (!this.cartToken) return { error: 'No cart token.' }
    const result = await this.apiGet('/api/cart/totals', { cartId: this.cartToken })
    if (result?.code === 200) {
      const r = result.result
      return {
        subtotal: r.subtotal,
        grand_total: r.grand_total,
        items_qty: r.items_qty,
        items: (r.items || []).map((i: any) => ({
          name: i.name,
          qty: i.qty,
          price: i.price,
          row_total: i.row_total,
        })),
      }
    }
    return { error: result?.result || 'Failed' }
  }

  // ── Timeslots ───────────────────────────────────────────────────────────

  DELIVERY_METHOD_ID = 419

  async getTimeslots(methodId?: number): Promise<any> {
    if (!this.cartToken) return { error: 'No cart token.' }
    return this.apiGet(
      `/api/ext/timeslot/get-method-data/${methodId || this.DELIVERY_METHOD_ID}/0`,
      { cartId: this.cartToken },
    )
  }

  async reserveTimeslot(timeslotId: number, date: string, email?: string, telephone?: string): Promise<any> {
    if (!this.cartToken) return { error: 'No cart token.' }
    const p = new URLSearchParams({ cartId: this.cartToken, storeCode: STORE })
    const r = await fetch(`${BASE}/api/ext/timeslot/reserve?${p}`, {
      method: 'POST',
      headers: this.authHeaders(),
      body: JSON.stringify({
        timeslotId,
        date,
        stationId: '',
        email: email || '',
        telephone: telephone || '',
      }),
    })
    return r.json()
  }

  // ── Pantry ──────────────────────────────────────────────────────────────

  loadPantry(): PantryItem[] {
    if (!existsSync(PANTRY_FILE)) return []
    try {
      const data = JSON.parse(readFileSync(PANTRY_FILE, 'utf8'))
      // Support both [{name}] and {items: [{name, quantity, unit}]} formats
      const items = Array.isArray(data) ? data : (data.items || [])
      return items
    } catch {
      return []
    }
  }

  savePantry(items: PantryItem[]): void {
    writeFileSync(PANTRY_FILE, JSON.stringify(items, null, 2))
  }

  pantryAdd(name: string, quantity?: number, unit?: string, category?: string): PantryItem[] {
    const items = this.loadPantry()
    const existing = items.find(i => i.name.toLowerCase() === name.toLowerCase())
    if (existing) {
      if (quantity && existing.quantity) existing.quantity += quantity
    } else {
      const entry: PantryItem = { name }
      if (quantity) entry.quantity = quantity
      if (unit) entry.unit = unit
      if (category) entry.category = category
      items.push(entry)
    }
    this.savePantry(items)
    return items
  }

  pantryRemove(name: string, quantity?: number): PantryItem[] {
    let items = this.loadPantry()
    const idx = items.findIndex(i => i.name.toLowerCase().includes(name.toLowerCase()))
    if (idx >= 0) {
      if (quantity && items[idx].quantity && items[idx].quantity! > quantity) {
        items[idx].quantity! -= quantity
      } else {
        items.splice(idx, 1)
      }
    }
    this.savePantry(items)
    return items
  }

  // ── Config ──────────────────────────────────────────────────────────────

  loadConfig(): any {
    if (!existsSync(CONFIG_FILE)) return {}
    try {
      return JSON.parse(readFileSync(CONFIG_FILE, 'utf8'))
    } catch {
      return {}
    }
  }
}
